crates/ra_project_model/src/lib.rs

```diff
@@ -406,19 +426,53 @@ fn find_rust_project_json(path: &Path) -> Option<PathBuf> {
     None
 }
 
-fn find_cargo_toml(path: &Path) -> Result<PathBuf> {
-    if path.ends_with("Cargo.toml") {
-        return Ok(path.to_path_buf());
-    }
+fn find_cargo_toml_in_parent_dir(path: &Path) -> Option<PathBuf> {
     let mut curr = Some(path);
     while let Some(path) = curr {
         let candidate = path.join("Cargo.toml");
         if candidate.exists() {
-            return Ok(candidate);
+            return Some(candidate);
         }
         curr = path.parent();
     }
-    Err(CargoTomlNotFoundError(path.to_path_buf()).into())
+
+    None
+}
+
+fn find_cargo_toml_in_child_dir(entities: ReadDir) -> Vec<PathBuf> {
+    // Only one level down to avoid cycles the easy way and stop a runaway scan with large projects
+    let mut valid_canditates = vec![];
+    for entity in entities.filter_map(Result::ok) {
+        let candidate = entity.path().join("Cargo.toml");
+        if candidate.exists() {
+            valid_canditates.push(candidate)
+        }
+    }
+    valid_canditates
+}
+
+fn find_cargo_toml(path: &Path) -> Result<PathBuf> {
+    let path_as_buf = path.to_path_buf();
+
+    if path.ends_with("Cargo.toml") {
+        return Ok(path.to_path_buf());
+    }
+
+    if let Some(p) = find_cargo_toml_in_parent_dir(path) {
+        return Ok(p);
+    }
+
+    let entities = match read_dir(path) {
+        Ok(entities) => entities,
+        Err(e) => return Err(CargoTomlSearchFileSystemError(path_as_buf, e.to_string()).into()),
```

Comment:
NOTE: This isn't even actually being too paranoid, as this is reasonably likely to happen in a workflow that utilizes something like docker for a dev environment.

If for example some output (ex. npm-s node_modules, cargo-s target) is generated by some process with rights that the rust-analyzer user cannot read then this would error and be hard to debug.